daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr0.scala

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package speedy
+
+/** SExpr0 -- AST for the speedy compiler pipeline.
+  *
+  * These are *not* the expression forms which run on the speedy machine. See SExpr.
+  *
+  * These are the expression forms which exist during the speedy compiler pipeline:
+  * The pipeline stages are:
+  *
+  * 1: convert from LF
+  *     - reducing binding forms (update & scenario becoming builtins)
+  *     - moving to de Bruijn indexing for variable
+  *     - moving to multi-argument applications and abstractions.
+  *
+  * 2: closure conversion
+  * 3: transform to ANF
+  *
+  * Stages 1 and 2 are in Compiler.scala; stage 3 in Anf.scala.
+  * During Stage3 (ANF transformation), we move from this type (SExpr0) to SExpr,
+  * and so have the expression form suitable for execution on a speedy machine.
+  */
+
+import com.daml.lf.data.Ref._
+import com.daml.lf.language.Ast
+import com.daml.lf.value.{Value => V}
+import com.daml.lf.speedy.SValue._
+import com.daml.lf.speedy.{SExpr => t}
+
+@SuppressWarnings(Array("org.wartremover.warts.Any"))
+object SExpr0 {
+
+  sealed abstract class SExpr extends Product with Serializable
+
+  sealed abstract class SExprAtomic extends SExpr
```

Comment:
```suggestion

sealed abstract class SExpr0 extends Product with Serializable

object SExpr0 {
  
  sealed abstract class SEAtomic extends SExpr0
  
    /** Reference to a value. On first lookup the evaluated expression is
    * stored in 'cached'.
    */
  final case class SEVal(ref: t.SDefinitionRef) extends SExpr
```

Reply:
1. Pushing back on this. Please see my free-standing comment at the bottom of the PR.
2. I added private[speedy] to `SExpr0`. Thank you for spotting this oversight